### PR TITLE
server: Clean up device_collection_change_callbacks registration on Drop.

### PR DESCRIPTION
**Pull request by kinetiknz**
If a CubebServer is ever dropped as a result of an error (e.g. client abruptly disconnects), any live device collection callbacks registered for that server must be removed.

### CODE REVIEW

File: server/src/server.rs [hunk 1]
```diff
@@ -349,6 +373,27 @@ pub struct CubebServer {
     shm_area_size: usize,
 }
 
+impl Drop for CubebServer {
+    fn drop(&mut self) {
+        if let Some(device_collection_change_callbacks) = &self.device_collection_change_callbacks {
+            debug!("CubebServer: dropped with device_collection_change_callbacks registered");
+            CONTEXT_KEY.with(|k| {
+                let mut state = k.borrow_mut();
+                if let Some(CubebContextState {
```
Comment (ChunMinChang) [inline server/src/server.rs:382]: Is it possible that `state` is `None` here? Doesn't that mean `with_local_context` is never called but we have a `CubebServer`?

File: server/src/server.rs [hunk 2]
```diff
@@ -47,14 +47,14 @@ impl CubebDeviceCollectionManager {
...
-            self.internal_register(context, true)?;
-        }
         server.borrow_mut().devtype.insert(devtype);
+        let mut servers = self.servers.lock().unwrap();
         if !servers.iter().any(|s| Rc::ptr_eq(s, server)) {
             servers.push(server.clone());
         }
+        if servers.len() == 1 {
+            self.internal_register(context, true)?;
+        }
```
Comment (ChunMinChang) [inline server/src/server.rs:57]: I don't think I understand why `servers.is_empty()` is changed to `servers.len() == 1`? Would you mind explaining it a bit?

**Review by ChunMinChang — COMMENTED** [review]
LGTM, hold r+ to wait for some answers

File: server/src/server.rs [hunk 3]
```diff
@@ -157,8 +180,9 @@ impl DevIdMap {
 }
 
 struct CubebContextState {
-    context: cubeb::Result<cubeb::Context>,
+    // `manager` must be dropped before the `context` is destroyed.
     manager: CubebDeviceCollectionManager,
+    context: cubeb::Result<cubeb::Context>,
```
[the remark this answers is not in the document]
Comment (kinetiknz) [reply server/src/server.rs:185]: That's right.

File: server/src/server.rs [hunk 4]
```diff
@@ -182,11 +206,11 @@ where
         let mut state = k.borrow_mut();
         if state.is_none() {
             *state = Some(CubebContextState {
-                context: cubeb_init_from_context_params(),
                 manager: CubebDeviceCollectionManager::new(),
+                context: cubeb_init_from_context_params(),
```
[the remark this answers is not in the document]
Comment (kinetiknz) [reply server/src/server.rs:210]: Thanks. I only made this change to align with the declaration order for consistency.

File: server/src/server.rs (hunk 1 above)
Comment (kinetiknz) [reply server/src/server.rs:382]: `with_local_context` is called in `CubebServer::process`, lazily initializing libcubeb on the first call.   Any time initialization fails to reach that point, the `CONTEXT_KEY` `state` will be `None` (e.g. if `bind_server` failed, or the `CubebServer` was disconnected before receiving a message).

File: server/src/server.rs (hunk 2 above)
Comment (kinetiknz) [reply server/src/server.rs:57]: Thanks for querying this, the logic is slightly wrong now.  I'll fix it up.

**Review by ChunMinChang — APPROVED** [review]

**Pull request merged**